jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/FrameFlusher.java

```diff
@@ -511,27 +518,62 @@ public String toString()
         return String.format("%s[queueSize=%d,aggregate=%s]",
             super.toString(),
             getQueueSize(),
-            batchBuffer);
+            _batchBuffer);
     }
 
-    private static class Entry extends FrameEntry
+    private class Entry extends FrameEntry implements CyclicTimeouts.Expirable
     {
-        private final long timeOfCreation = System.currentTimeMillis();
+        private long _expiry = Long.MAX_VALUE;
 
         private Entry(Frame frame, Callback callback, boolean batch)
         {
             super(frame, callback, batch);
+
+            long currentTime = NanoTime.now();
+            if (_frameTimeout > 0)
+            {
+                long frameExpiry = Math.addExact(currentTime, Duration.ofMillis(_frameTimeout).toNanos());
```

Comment:
I think this avoids the allocation of a Duration object
```suggestion
                long frameExpiry = Math.addExact(currentTime, TimeUnit.MILLISECONDS.toNanos(_frameTimeout));
```

Use TimeUnit instead of Duration below as well.